anaconda_packaging_utils/api/jira_api.py

```diff
@@ -0,0 +1,76 @@
+File:           jira_api.py
+Description:    Wrapper library that provides tools for using the Python JIRA API. The primary purpose of this wrapper
+                is to simplify and standardize the authentication process.
+
+                JIRA API links:
+                  - Source: https://github.com/pycontribs/jira
+                  - Docs: https://jira.readthedocs.io/index.html#
+"""
+import logging
+from typing import Final
+
+from jira.client import JIRA
+
+from anaconda_packaging_utils.storage.config_data import ConfigData
+
+# Logging object for this module
+log = logging.getLogger(__name__)
+
+# Where our JIRA boards are hosted
+_JIRA_HOST_URL: Final[str] = "https://anaconda.atlassian.net/"
+
+
+class ApiException(Exception):
+    """
+    Generic exception indicating an unrecoverable failure of this API.
+
+    This exception is meant to condense many possible failures into one generic error. The thinking is, if the calling
+    code runs into any API failure, there isn't much that can be done. So it is easier for the caller to handle one
+    exception than many exception types.
+    """
+
+    def __init__(self, message: str):
+        """
+        Constructs an API exception
+        :param message: String description of the issue encountered.
+        """
+        super().__init__(message if len(message) else "An unknown API issue was encountered.")
+
+
+class JiraApi:
+    """
+    Singleton wrapper to the Python Jira project. This "ensures" that we only construct and authenticate the underlying
+    Jira object once.
+    """
+
+    # The Jira API is wrapped in a list as a cheesy way to work around an initialization problem. Defaulting to `None`
+    # or using an `Optional` causes the static analyzer to freak out on every use of the `__jira`, even if the static
+    # variable has to have been initialized by instance-method call time.
+    __jira: list[JIRA] = []
+
+    def __init__(self) -> None:
+        """
+        Constructs a JiraApi instance
+        :raises ApiException: If there was a failure to authenticate.
+        """
+        if len(JiraApi.__jira) == 0:
+            data_store: Final[ConfigData] = ConfigData()
+            try:
+                JiraApi.__jira.append(
+                    JIRA(
+                        _JIRA_HOST_URL,
+                        basic_auth=(data_store["user_info.email"], data_store["token.jira"]),
+                    )
+                )
+            except Exception as e:
+                raise ApiException("Failed to auth or connect to JIRA") from e
+
+    def get_jira(self) -> JIRA:
+        """
+        Exposes an authenticated JIRA API instance directly to the caller, allowing for full use of the API.
+        As this is a member function, successful construction of a `JiraApi` instance must have occurred previously
+        for this to be able to be called.
+        :returns: Authenticated instance of the underlying JIRA API
+        """
+        return JiraApi.__jira[0]
```

Comment:
not thread safe with singleton initialization but probably not mean to be thread safe

Reply:
JC and I briefly talked about this. My initial callback mechanism I just redacted could have leant itself to being more thread safe. That being said, it's unclear (from the small amount of research I've done) if the underlying JIRA API or the GitHub API are threadsafe anyways.

Reply:
The problem with singleton can start be apparent when you start doing test, you returning JIRA object which would hold its state across different test cases, if you call any stateful method on JIRA object the result could be observer in follow up tests, which could break the 1st rule that the tests should be run independent from each other. The solution for ( if you still want to use singleton ) is to have a wrapper singleton in your test code which would be able to reset the underlying instance each time new test is run